### PR TITLE
Fixed #27147 -- Allowed specifying bounds of tuple inputs for non-discrete range fields.

### PR DESCRIPTION
**Pull request by gmcrocetti**
This PR fixes [#27147](https://code.djangoproject.com/ticket/27147).

I could think of two UIs to fix the stated problem:

1. We can add a `default_bounds` as an argument to `RangeField` class, and thus all subclasses. The problem of this approach is that we may add some cutler for discrete ranges. For example, IMO it makes no sense to set a `default_bounds` value for a `IntegerRangeField` as it'll always be converted to `[)`, cluttering user's experience. Please, let me know if I'm missing something.
2. We can specialize `RangeField` for continuous ranges. With this approach we're preventing misuse.

I implemented approach 2.

I'm following a conservative approach for `default_bounds`. For example:

```python
from psycopg2.extras import DateTimeTZRange

lower_date = datetime.date(2014, 1, 1)
upper_date = datetime.date(2014, 2, 2)

class DummyRange(models.Model):
        timestamps = DateTimeRangeField(blank=True, null=True, default_bounds='[]')

DummyRange.objects.create(timestamps=DateTimeTZRange(lower_date, upper_date))  # saved with bounds='[)'
DummyRange.objects.create(timestamps=DateTimeTZRange(lower_date, upper_date, '()'))  # saved with bounds='()'
DummyRange.objects.create(timestamps=(lower_date, upper_date))  # saved with bounds='[]' - default_bounds
```
The reasoning behind is that a user knows what's he's doing if he created a `Range` object (Adding a validator sounds like a good idea). I still need to document whichever behavior we choose.

Looking forward for feedback before finishing this feature.

### CODE REVIEW

File: django/contrib/postgres/fields/ranges.py [hunk 1]
```diff
@@ -92,25 +95,62 @@ def set_attributes_from_name(self, name):
...
     def value_to_string(self, obj):
         value = self.value_from_object(obj)
+        base_field = self.base_field
+
         if value is None:
             return None
+
+        if isinstance(value, (list, tuple)):
+            bounds = getattr(self, 'default_bounds', CANONICAL_RANGE_BOUNDS)
+            value = self.range_type(value[0], value[1], bounds=bounds)
```
Comment (gmcrocetti) [inline django/contrib/postgres/fields/ranges.py:105]: This change is not related with the PR. Serialization wasn't working for models with list or tuples for range fields.

**Comment by gmcrocetti** [comment]
I'd appreciate any enlightenment on why CI is failing. `tox -e py39-postgres -- --settings=tests.test_pg` is running without errors while `tox` is failing, both on my machine.

**Comment by charettes** [comment]
@gmcrocetti not sure you got tox passing tests locally but the error on Jenkins is the following

https://djangoci.com/job/pull-requests-bionic/database=sqlite3,label=bionic-pr,python=python3.8/12028/console

```
Traceback (most recent call last):
  File "./runtests.py", line 646, in <module>
    failures = django_tests(
  File "./runtests.py", line 364, in django_tests
    test_labels, state = setup_run_tests(verbosity, start_at, start_after, test_labels)
  File "./runtests.py", line 298, in setup_run_tests
    apps.set_installed_apps(settings.INSTALLED_APPS)
  File "/home/jenkins/workspace/pull-requests-bionic/database/sqlite3/label/bionic-pr/python/python3.8/django/apps/registry.py", line 355, in set_installed_apps
    self.populate(installed)
  File "/home/jenkins/workspace/pull-requests-bionic/database/sqlite3/label/bionic-pr/python/python3.8/django/apps/registry.py", line 114, in populate
    app_config.import_models()
  File "/home/jenkins/workspace/pull-requests-bionic/database/sqlite3/label/bionic-pr/python/python3.8/django/apps/config.py", line 300, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/jenkins/workspace/pull-requests-bionic/database/sqlite3/label/bionic-pr/python/python3.8/tests/postgres_tests/models.py", line 132, in <module>
    class RangesModel(PostgreSQLModel):
  File "/home/jenkins/workspace/pull-requests-bionic/database/sqlite3/label/bionic-pr/python/python3.8/tests/postgres_tests/models.py", line 138, in RangesModel
    timestamps_with_closed_bounds = DateTimeRangeField(blank=True, null=True, default_bounds='[]')
TypeError: __init__() got an unexpected keyword argument 'default_bounds'
```

**Comment by gmcrocetti** [comment]
Yes, `tox` wasn't working as I said. It was working **only** for the postgres environment. Nonetheless the problem was a missing DummyField with `default_bounds` for non postgres setups.

**Comment by charettes** [comment]
@gmcrocetti I've seen that you managed to get things working but for the future https://github.com/django/django-docker-box is usually the easiest way to get test running against different backends locally.

**Comment by gmcrocetti** [comment]
@charettes . I'm not familiar with the django's revision queue. Is there anything else expected from me or you folks handle it now ?

**Comment by charettes** [comment]
@gmcrocetti the ticket still has the _needs documentation_ checkbox ticked hence why it doesn't show up in the review queue.

Since there's still unanswered questions about the approach to take I'd try to gather feedback from the reporter as well as the folks involved in #10484. If that doesn't yield any result I'd post to [the developer mailing](https://docs.djangoproject.com/en/3.2/internals/mailing-lists/#django-developers) list to reach a wider audience.

**Comment by gmcrocetti** [comment]
@jdufresne , would you like taking a view ?

File: tests/postgres_tests/test_ranges.py [hunk 2]
```diff
@@ -50,6 +50,39 @@ class Model(PostgreSQLModel):
...
+    def test_discrete_range_fields_invalid_default_bounds_argument(self):
+        discrete_range_fields = [
+            pg_fields.IntegerRangeField,
+            pg_fields.BigIntegerRangeField,
+            pg_fields.DateRangeField
+        ]
+
+        for discrete_range_field in discrete_range_fields:
+            field_class_name = f'{discrete_range_field.__class__.__name__}'
+            with self.assertRaises(TypeError, msg=f"Cannot use 'default_bounds' with {field_class_name}"):
```
Comment (gmcrocetti) [inline tests/postgres_tests/test_ranges.py:62]: I'm used with pytest parametrization for such cases. Is there a recommend code-style regarding that ?

File: django/contrib/postgres/forms/ranges.py [hunk 3]
```diff
@@ -42,6 +42,7 @@ def __init__(self, **kwargs):
             kwargs['fields'] = [self.base_field(required=False), self.base_field(required=False)]
         kwargs.setdefault('required', False)
         kwargs.setdefault('require_all_fields', False)
+        self.default_bounds = kwargs.pop('default_bounds', '[)')
```
[the remark this answers is not in the document]
Comment (gmcrocetti) [reply django/contrib/postgres/forms/ranges.py]: This is way more elegant. :+1:

File: django/contrib/postgres/fields/ranges.py [hunk 4]
```diff
@@ -95,9 +102,9 @@ def value_to_string(self, obj):
         if value is None:
             return None
         if value.isempty:
-            return json.dumps({"empty": True})
+            return json.dumps({'empty': True})
         base_field = self.base_field
-        result = {"bounds": value._bounds}
+        result = {'bounds': value._bounds}
```
Comment (felixxm) [inline django/contrib/postgres/fields/ranges.py:107]: Please revert unrelated changes.

File: django/contrib/postgres/fields/ranges.py [hunk 5]
```diff
@@ -95,9 +102,9 @@ def value_to_string(self, obj):
         if value is None:
             return None
         if value.isempty:
-            return json.dumps({"empty": True})
+            return json.dumps({'empty': True})
```
Comment (felixxm) [inline django/contrib/postgres/fields/ranges.py:105]: Please revert unrelated changes.

File: django/contrib/postgres/fields/ranges.py [hunk 6]
```diff
@@ -47,6 +50,10 @@ def __init__(self, *args, **kwargs):
         # Initializing base_field here ensures that its model matches the model for self.
         if hasattr(self, 'base_field'):
             self.base_field = self.base_field()
+
+        if 'default_bounds' in kwargs:
+            raise TypeError(f"Cannot use 'default_bounds' with {self.__class__.__name__}.")
+
```
Comment (felixxm) [inline django/contrib/postgres/fields/ranges.py:56]: I'd move this at the beginning of `__init__()`. Also, chop blank lines.

File: tests/postgres_tests/test_ranges.py (hunk 2 above)
Comment (felixxm) [reply tests/postgres_tests/test_ranges.py:62]:
```suggestion
            msg = f"Cannot use 'default_bounds' with {field_class_name}."
            with self.assertRaisesMessage(TypeError, msg):
```

**Comment by gmcrocetti** [comment]
@felixxm . Whos responsible for writing the documentation ?

File: django/contrib/postgres/fields/ranges.py [hunk 7]
```diff
@@ -112,6 +117,34 @@ def formfield(self, **kwargs):
...
+    """
+    This class specializes RangeField allowing user to choose a default bounds value
+    for the field. This assumption is conservative in the sense that we enforce the
+    chosen 'default_bounds' exclusively for inputs of the types (list, tuple).
+    We're not replacing the bounds attribute by 'default_bounds' for objects of the type
+    `psycopg2.extras.Range`.
+    """
+
+    def __init__(self, *args, default_bounds=CANONICAL_RANGE_BOUNDS, **kwargs):
+        self.default_bounds = default_bounds
```
Comment (gmcrocetti) [inline django/contrib/postgres/fields/ranges.py:130]: Should we preemptively validate `default_bounds` to be one of the four possible values ?  Misuse may cause errors down the stack.

**Comment by gmcrocetti** [comment]
@felixxm , pushed documentation and left one question regarding the validation of `default_bounds`.
Great thanks

File: django/contrib/postgres/fields/ranges.py (hunk 7 above)
Comment (felixxm) [reply django/contrib/postgres/fields/ranges.py:130]: Yes, we should raise a `ValueError` for incorrect values.

Comment (gmcrocetti) [reply django/contrib/postgres/fields/ranges.py:132]: Validation was added.

**Review by felixxm — COMMENTED** [review]
@gmcrocetti Thanks for updates :+1: 

We should also add release notes (`4.1.txt`) and update [Range Fields description](https://docs.djangoproject.com/en/3.2/ref/contrib/postgres/fields/#range-fields).

File: django/contrib/postgres/fields/ranges.py [hunk 8]
```diff
@@ -112,6 +117,36 @@ def formfield(self, **kwargs):
...
+            return self.range_type(value[0], value[1], self.default_bounds)
+        return super().get_prep_value(value)
+
+    def formfield(self, **kwargs):
+        kwargs.setdefault('default_bounds', self.default_bounds)
+        return super().formfield(**kwargs)
+
+    def deconstruct(self):
+        name, path, args, kwargs = super().deconstruct()
+        kwargs['default_bounds'] = self.default_bounds
```
[the remark this answers is not in the document]
Comment (gmcrocetti) [reply django/contrib/postgres/fields/ranges.py:146]: Yeah, makes sense.
Checking for `self.default_bounds != '[)'` isn't enough  as `default_bounds` is always set ?

File: django/contrib/postgres/fields/ranges.py [hunk 9]
```diff
@@ -112,6 +117,36 @@ def formfield(self, **kwargs):
...
+class ContinuousRangeField(RangeField):
+    """
+    This class specializes RangeField allowing user to choose a default bounds value
+    for the field. This assumption is conservative in the sense that we enforce the
+    chosen 'default_bounds' exclusively for inputs of the types (list, tuple).
+    We're not replacing the bounds attribute by 'default_bounds' for objects of the type
+    `psycopg2.extras.Range`.
+    """
+
+    def __init__(self, *args, default_bounds=CANONICAL_RANGE_BOUNDS, **kwargs):
```
[the remark this answers is not in the document]
Comment (gmcrocetti) [reply django/contrib/postgres/fields/ranges.py:129]: I don't have a strong opinion, but we're going to use it in two places: `__init__` and `deconstruct`. Do you still believe we should move to the "simple string" approach ?

File: docs/ref/contrib/postgres/fields.txt [hunk 10]
```diff
@@ -538,23 +538,39 @@ excluded, that is ``[)`` (see the PostgreSQL documentation for details about
...
+.. class:: DecimalRangeField(default_bounds='[)', **options)
 
     Stores a range of floating point values. Based on a
     :class:`~django.db.models.DecimalField`. Represented by a ``numrange`` in
     the database and a :class:`~psycopg2:psycopg2.extras.NumericRange` in
     Python.
 
+    .. attribute:: DecimalRangeField.default_bounds
+
+        Optional. The value for ``bounds`` at tuple inputs.
```
[the remark this answers is not in the document]
Comment (gmcrocetti) [reply docs/ref/contrib/postgres/fields.txt:550]: Definitely looks better ! :). Thanks

**Comment by gmcrocetti** [comment]
@felixxm , updated.

File: django/contrib/postgres/fields/ranges.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply django/contrib/postgres/fields/ranges.py]: It's an attribute so folks can try to change it dynamically. I would add `if self.default_bounds and ..`

[the remark this answers is not in the document]
Comment (gmcrocetti) [reply django/contrib/postgres/fields/ranges.py]: Ok, thanks for letting me know. Defensive programming in the first place ! :)

**Review by felixxm — COMMENTED** [review]
@gmcrocetti Thanks for updates :+1: Welcome aboard :boat: 

I pushed final edits to docs and tests.

File: tests/postgres_tests/test_apps.py [hunk 11]
```diff
@@ -58,3 +60,41 @@ def assertNotSerializable():
...
+                        field_string = f'{field_string}(default=psycopg2.extras.{default!r})'
+
+                    serialized_field, imports = MigrationWriter.serialize(field)
+                    self.assertEqual(imports, {
+                        'import django.contrib.postgres.fields.ranges',
+                        'import psycopg2.extras',
+                    })
+                    self.assertIn(field_string, serialized_field)
+
+        assertNotSerializable()
```
[the remark this answers is not in the document]
Comment (gmcrocetti) [reply tests/postgres_tests/test_apps.py]: Alright, makes sense. Thanks a lot for the refactoring !!

**Comment by gmcrocetti** [comment]
@felixxm , thanks a lot for all the effort. Documentation is pretty clear and thanks for the tests. Just left one `nit`.
Thanks again ! :)

File: tests/postgres_tests/test_ranges.py [hunk 12]
```diff
@@ -478,6 +536,8 @@ class TestSerialization(PostgreSQLSimpleTestCase):
         '"bigints": null, "timestamps": "{\\"upper\\": \\"2014-02-02T12:12:12+00:00\\", '
         '\\"lower\\": \\"2014-01-01T00:00:00+00:00\\", \\"bounds\\": \\"[)\\"}", '
         '"timestamps_inner": null, '
+        '"timestamps_closed_bounds": "{\\"upper\\": \\"2014-02-02T12:12:12+00:00\\", '
```
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/postgres_tests/test_ranges.py:539]: `timestamps_closed_bounds` field is defined with closed bounds in `default_bounds` so the current name is good, IMO. This test is to ensure that `default_bounds` doesn't impact open bounds pass in `DateTimeTZRange`.

[the remark this answers is not in the document]
Comment (gmcrocetti) [reply tests/postgres_tests/test_ranges.py:539]: Humm...I see it. Sounds good !
